common/psbt_open.c

```diff
@@ -0,0 +1,389 @@
+			continue;
+		}
+
+		result = compare_serials(&orig->outputs[i].unknowns,
+					 &new->outputs[j].unknowns);
+		if (result == -1) {
+			ADD(output, rm_outs, orig, i);
+			i++;
+			continue;
+		}
+		if (result == 1) {
+			ADD(output, added_outs, new, j);
+			j++;
+			continue;
+		}
+		if (!output_identical(orig, i, new, j)) {
+			ADD(output, rm_outs, orig, i);
+			ADD(output, added_outs, new, j);
+		}
+		i++;
+		j++;
+	}
+
+	return tal_count(*added_ins) != 0 ||
+		tal_count(*rm_ins) != 0 ||
+		tal_count(*added_outs) != 0 ||
+		tal_count(*rm_outs) != 0;
+}
+
+void psbt_input_add_serial_id(struct wally_psbt_input *input,
+			      u16 serial_id)
+{
+	u8 *key = psbt_make_key(tmpctx, PSBT_TYPE_SERIAL_ID, NULL);
+	psbt_input_add_unknown(input, key, &serial_id, sizeof(serial_id));
+}
+
+
+void psbt_output_add_serial_id(struct wally_psbt_output *output,
+			       u16 serial_id)
+{
+	u8 *key = psbt_make_key(tmpctx, PSBT_TYPE_SERIAL_ID, NULL);
+	psbt_output_add_unknown(output, key, &serial_id, sizeof(serial_id));
+}
+
+bool psbt_has_serial_input(struct wally_psbt *psbt, u16 serial_id)
+{
+	for (size_t i = 0; i < psbt->num_inputs; i++) {
+		u16 in_serial;
+		if (!psbt_get_serial_id(&psbt->inputs[i].unknowns, &in_serial))
+			continue;
+		if (in_serial == serial_id)
+			return true;
+	}
+	return false;
+}
+
+bool psbt_has_serial_output(struct wally_psbt *psbt, u16 serial_id)
+{
+	for (size_t i = 0; i < psbt->num_outputs; i++) {
+		u16 out_serial;
+		if (!psbt_get_serial_id(&psbt->outputs[i].unknowns, &out_serial))
+			continue;
+		if (out_serial == serial_id)
+			return true;
+	}
+	return false;
+}
+
+void psbt_input_add_max_witness_len(struct wally_psbt_input *input,
+				    u16 max_witness_len)
+{
+	u8 *key = psbt_make_key(NULL, PSBT_TYPE_MAX_WITNESS_LEN, NULL);
+	psbt_input_add_unknown(input, key, &max_witness_len, sizeof(max_witness_len));
+	tal_free(key);
+}
+
+
+bool psbt_input_get_max_witness_len(struct wally_psbt_input *input,
+				    u16 *max_witness_len)
+{
+	size_t value_len;
+	void *result = psbt_get_lightning(&input->unknowns,
+					  PSBT_TYPE_MAX_WITNESS_LEN,
+					  &value_len);
+	if (!result)
+		return false;
+
+	if (value_len != sizeof(*max_witness_len))
+		return false;
+
+	memcpy(max_witness_len, result, value_len);
+	return true;
+}
+
+bool psbt_has_required_fields(struct wally_psbt *psbt)
+{
+	u16 max_witness, serial_id;
+	for (size_t i = 0; i < psbt->num_inputs; i++) {
+		struct wally_psbt_input *input = &psbt->inputs[i];
+
+		if (!psbt_get_serial_id(&input->unknowns, &serial_id))
+			return false;
+
+		/* Inputs had also better have their max_witness_lens
+		 * filled in! */
+		if (!psbt_input_get_max_witness_len(input, &max_witness))
+			return false;
+
+		/* Required because we send the full tx over the wire now */
+		if (!input->utxo)
+			return false;
+
+		/* If is P2SH, redeemscript must be present */
+		const u8 *outscript =
+			wally_tx_output_get_script(tmpctx,
+				&input->utxo->outputs[psbt->tx->inputs[i].index]);
```

Comment:
I think you want to assert this is in bounds for the utxo outputs before dereferencing here, just to be sure.

Reply:
ack, updated